### ISSUE TITLE
ptr::read and ptr::write (and others?) should specify alignment requirements in their contract

### ISSUE THREAD

**Issue by whitequark** [rust-lang/rust#36450]
Currently, these functions do not say anything about whether the raw pointer has to be aligned. They are also currently implemented with copy_nonoverlapping, which means they are safe to use with unaligned data. However, @ubsan tells me that this is an implementation detail that cannot be relied on.

In any case, these functions should explicitly specify their alignment requirements. Similarly, ptr::copy{,_nonoverlapping} also should do that, since right now they do not explicitly state that in their contract either. (They do reference C's memcpy and memmove, but I don't think that a C reference is a good way to define normative semantics.)

**Comment by ubsan** [comment]
ping @Amanieu

**Comment by Amanieu** [comment]
See https://github.com/rust-lang/rfcs/pull/1725

**Comment by Amanieu** [comment]
Also keep in mind that `copy(_nonoverlapping)` *does* have alignment requirements: it's based on the type of the pointer. So if you pass a `*mut u32` to `copy_nonoverlapping` then LLVM will assume that the pointer is aligned to 4 bytes.

**Comment by whitequark** [comment]
That then should be explicitly specified also, because it is *definitely* not a part of the C memcpy semantics.

**Comment by bluss** [comment]
Maybe it's obvious in this discussion, but `*ptr` (copy to value from raw pointer) requires ptr to be aligned, of course. With that as the baseline, I think it would have to be assumed that read/write use that rule if nothing else is documented -- but why not document it explicitly.

**Comment by ubsan** [comment]
@bluss That's also never documented, to be clear.

**Comment by tbu-** [comment]
Seems to be fixed, the documentation says:
>The pointer must be aligned; use `write_unaligned` if that is not the case.

and similarly for `ptr::read`.

**Comment by ubsan** [comment]
Recommend closing.

**Comment by whitequark** [comment]
copy{,_nonoverlapping} don't have mention of alignment in the documentation still.

**Comment by arielb1** [comment]
@whitequark

That's because it has none.

**Comment by whitequark** [comment]
> Also keep in mind that copy(_nonoverlapping) does have alignment requirements: it's based on the type of the pointer. So if you pass a *mut u32 to copy_nonoverlapping then LLVM will assume that the pointer is aligned to 4 bytes.

@Amanieu above disagrees...

**Comment by tbu-** [comment]
The linked RFC (by @Amanieu) states that it uses `copy_nonoverlapping` for `{read,write}_unaligned`.

**Comment by arielb1** [comment]
@whitequark

I had the mistaken impression that `copy_nonoverlapping` took a u8 parameter.

**Comment by Mark-Simulacrum** [comment]
Okay, so this appears to be a documentation issue -- we need to update `ptr::copy{,_nonoverlapping}` to state something about alignment.

**Comment by QuietMisdreavus** [comment]
The docs team discussed this today. Here's the current state of the `ptr` module, with respect to alignment:

    <ubsan> because anything in the standard library which takes a raw pointer assumes validity of the
            raw pointer, unless it's explicitly denoted `_unaligned`, in which case size must be valid,
            but alignment might not be
    (...)
    <ubsan> copy{_nonoverlapping}, read_volatile, probably replace, probably swap (uses "valid",
            probably could be more wordy), write_volatile, swap_nonoverlapping
    <ubsan> we should also standardize this documentation across all of the functions
    (...)
    <ubsan> (some put their requirements under "Safety", some put it under "Undefined Behavior", etc.)

Basically, to really close this issue it would be prudent to go over every function in `ptr` and leave a note about the assumed alignment requirements of it.

Since https://github.com/rust-lang/rust/issues/29371 is still open, i'm considering this issue part of that one.

**Comment by ecstatic-morse** [comment]
@steveklabnik. This was completed in #53783. All functions in `std::ptr` now state their alignment requirements in the "Safety" section.

**Comment by steveklabnik** [comment]
Thank you!
